File: src/DynamixelSDK/python/src/dynamixel_easy_sdk/connector.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

################################################################################
# Copyright 2025 ROBOTIS CO., LTD.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

# Author: Hyungyu Kim

from typing import List
import serial

from dynamixel_easy_sdk.dynamixel_error import DxlError
from dynamixel_easy_sdk.dynamixel_error import DxlRuntimeError
from dynamixel_easy_sdk.group_executor import GroupExecutor
from dynamixel_easy_sdk.motor import Motor
from dynamixel_sdk import PacketHandler
from dynamixel_sdk import PortHandler


class Connector:

    PROTOCOL_VERSION = 2.0
    _packet_handler = None

    def __init__(self, port_name: str, baud_rate: int):
        self._port_handler = PortHandler(port_name)
        if Connector._packet_handler is None:
            Connector._packet_handler = PacketHandler(Connector.PROTOCOL_VERSION)

        try:
            if not self._port_handler.setBaudRate(baud_rate):
                raise DxlRuntimeError('Invalid baudrate specified')
        except serial.SerialException as e:
            errno = getattr(e, 'errno', None)
            if errno == 2:
                raise DxlRuntimeError(f'Port "{port_name}" does not exist') from e
            elif errno == 16:
                raise DxlRuntimeError(DxlError.SDK_COMM_PORT_BUSY) from e
            else:
                raise DxlRuntimeError('Failed to open port') from e
        except Exception as e:
            raise DxlRuntimeError('Failed to open port') from e

    def createMotor(self, motor_id: int) -> Motor:
        model_number = self.ping(motor_id)
        return Motor(motor_id, model_number, self)

    def createAllMotors(self, start_id: int = 0, end_id: int = 252) -> List[Motor]:
        if not (0 <= start_id <= 252 and 0 <= end_id <= 252 and start_id <= end_id):
            raise DxlRuntimeError('ID must be between 0 and 252, and start_id <= end_id')

        motor_ids = self.broadcastPing()
        motors = []
        for motor_id in motor_ids:
            if start_id <= motor_id <= end_id:
                motors.append(self.createMotor(motor_id))
        return motors

    def createGroupExecutor(self):
        return GroupExecutor(self)

    def _checkError(self, dxl_comm_result, dxl_error):
        if dxl_comm_result != DxlError.SDK_COMM_SUCCESS:
            raise DxlRuntimeError(DxlError(dxl_comm_result))
        if dxl_error != DxlError.SDK_COMM_SUCCESS:
            raise DxlRuntimeError(DxlError(dxl_error))

    def read1ByteData(self, motor_id: int, address: int) -> int:
        value, dxl_comm_result, dxl_error = Connector._packet_handler.read1ByteTxRx(
            self._port_handler,
            motor_id,
            address)
        self._checkError(dxl_comm_result, dxl_error)
        return value

    def read2ByteData(self, motor_id: int, address: int) -> int:
        value, dxl_comm_result, dxl_error = Connector._packet_handler.read2ByteTxRx(
            self._port_handler,
            motor_id,
            address)
        self._checkError(dxl_comm_result, dxl_error)
        return value

    def read4ByteData(self, motor_id: int, address: int) -> int:
        value, dxl_comm_result, dxl_error = Connector._packet_handler.read4ByteTxRx(
            self._port_handler,
            motor_id,
            address)
        self._checkError(dxl_comm_result, dxl_error)
        return value

    def write1ByteData(self, motor_id: int, address: int, value: int):
        dxl_comm_result, dxl_error = Connector._packet_handler.write1ByteTxRx(
            self._port_handler,
            motor_id,
            address,
            value)
        self._checkError(dxl_comm_result, dxl_error)

    def write2ByteData(self, motor_id: int, address: int, value: int):
        dxl_comm_result, dxl_error = Connector._packet_handler.write2ByteTxRx(
            self._port_handler,
            motor_id,
            address,
            value)
        self._checkError(dxl_comm_result, dxl_error)

    def write4ByteData(self, motor_id: int, address: int, value: int):
        dxl_comm_result, dxl_error = Connector._packet_handler.write4ByteTxRx(
            self._port_handler,
            motor_id,
            address,
            value)
        self._checkError(dxl_comm_result, dxl_error)

    def reboot(self, motor_id: int):
        dxl_comm_result, dxl_error = Connector._packet_handler.reboot(self._port_handler, motor_id)
        self._checkError(dxl_comm_result, dxl_error)

    def ping(self, motor_id: int) -> int:
        model_number, dxl_comm_result, dxl_error = Connector._packet_handler.ping(
            self._port_handler,
            motor_id)
        self._checkError(dxl_comm_result, dxl_error)
        return model_number

    def broadcastPing(self) -> List[int]:
        ids, dxl_comm_result = Connector._packet_handler.broadcastPing(self._port_handler)
        self._checkError(dxl_comm_result, DxlError.SDK_COMM_SUCCESS)
        return ids

    def factoryReset(self, motor_id: int, option: int):
        dxl_comm_result, dxl_error = Connector._packet_handler.factoryReset(
            self._port_handler,
            motor_id,
            option)
        self._checkError(dxl_comm_result, dxl_error)

    def closePort(self):
        self._port_handler.closePort()

File: src/DynamixelSDK/ros/dynamixel_sdk/src/dynamixel_easy_sdk/connector.cpp
// Copyright 2025 ROBOTIS CO., LTD.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Hyungyu Kim

#include "dynamixel_easy_sdk/connector.hpp"

static constexpr float PROTOCOL_VERSION = 2.0f;

namespace dynamixel
{
Connector::Connector(const std::string & port_name, int baud_rate)
{
  port_handler_ = std::unique_ptr<PortHandler>(PortHandler::getPortHandler(port_name.c_str()));
  packet_handler_ = PacketHandler::getPacketHandler(PROTOCOL_VERSION);

  if (!port_handler_->openPort()) {
    throw DxlRuntimeError("Failed to open the port!");
  }

  if (!port_handler_->setBaudRate(baud_rate)) {
    throw DxlRuntimeError("Failed to set the baud rate!");
  }
}

Connector::~Connector()
{
  port_handler_->closePort();
}

std::unique_ptr<Motor> Connector::createMotor(uint8_t id)
{
  Result<uint16_t, DxlError> result = ping(id);
  if (!result.isSuccess()) {
    throw DxlRuntimeError(getErrorMessage(result.error()));
  }
  return std::make_unique<Motor>(id, result.value(), this);
}

std::vector<std::unique_ptr<Motor>> Connector::createAllMotors(int start_id, int end_id)
{
  if (start_id < 0 || start_id > 252 || end_id < 0 || end_id > 252 || start_id > end_id) {
    throw DxlRuntimeError("Invalid ID range. ID should be in the range of 0 to 252.");
  }
  std::vector<std::unique_ptr<Motor>> motors;
  Result<std::vector<uint8_t>, DxlError> result = broadcastPing();
  if (!result.isSuccess()) {
    throw DxlRuntimeError(getErrorMessage(result.error()));
  }
  for (auto & id : result.value()) {
    if (id >= start_id && id <= end_id) {
      motors.push_back(createMotor(id));
    }
  }
  return motors;
}

std::unique_ptr<GroupExecutor> Connector::createGroupExecutor()
{
  return std::make_unique<GroupExecutor>(this);
}

Result<uint8_t, DxlError> Connector::read1ByteData(uint8_t id, uint16_t address)
{
  uint8_t dxl_error = 0;
  uint8_t data = 0;
  int dxl_comm_result = packet_handler_->read1ByteTxRx(
    port_handler_.get(),
    id, address,
    &data,
    &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return data;
}

Result<uint16_t, DxlError> Connector::read2ByteData(uint8_t id, uint16_t address)
{
  uint8_t dxl_error = 0;
  uint16_t data = 0;
  int dxl_comm_result = packet_handler_->read2ByteTxRx(
    port_handler_.get(),
    id, address,
    &data,
    &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return data;
}

Result<uint32_t, DxlError> Connector::read4ByteData(uint8_t id, uint16_t address)
{
  uint8_t dxl_error = 0;
  uint32_t data = 0;
  int dxl_comm_result = packet_handler_->read4ByteTxRx(
    port_handler_.get(),
    id, address,
    &data,
    &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return data;
}

Result<void, DxlError> Connector::write1ByteData(uint8_t id, uint16_t address, uint8_t value)
{
  uint8_t dxl_error = 0;
  int dxl_comm_result = packet_handler_->write1ByteTxRx(
    port_handler_.get(),
    id, address, value,
    &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return {};
}

Result<void, DxlError> Connector::write2ByteData(uint8_t id, uint16_t address, uint16_t value)
{
  uint8_t dxl_error = 0;
  int dxl_comm_result = packet_handler_->write2ByteTxRx(
    port_handler_.get(),
    id, address,
    value,
    &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return {};
}

Result<void, DxlError> Connector::write4ByteData(uint8_t id, uint16_t address, uint32_t value)
{
  uint8_t dxl_error = 0;
  int dxl_comm_result = packet_handler_->write4ByteTxRx(
    port_handler_.get(),
    id, address,
    value,
    &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return {};
}

Result<void, DxlError> Connector::reboot(uint8_t id)
{
  uint8_t dxl_error = 0;
  int dxl_comm_result = packet_handler_->reboot(port_handler_.get(), id, &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return {};
}

Result<uint16_t, DxlError> Connector::ping(uint8_t id)
{
  uint8_t dxl_error = 0;
  uint16_t data = 0;
  int dxl_comm_result = packet_handler_->ping(port_handler_.get(), id, &data, &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return data;
}

Result <std::vector<uint8_t>, DxlError> Connector::broadcastPing()
{
  std::vector<uint8_t> ids;
  int dxl_comm_result = packet_handler_->broadcastPing(port_handler_.get(), ids);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  return ids;
}

Result<void, DxlError> Connector::factoryReset(uint8_t id, uint8_t option)
{
  uint8_t dxl_error = 0;
  int dxl_comm_result = packet_handler_->factoryReset(port_handler_.get(), id, option, &dxl_error);
  if (dxl_comm_result != COMM_SUCCESS) {
    return static_cast<DxlError>(dxl_comm_result);
  }
  if (dxl_error != 0) {
    return static_cast<DxlError>(dxl_error);
  }
  return {};
}
}  // namespace dynamixel
